x-pack/agent/pkg/fleetapi/checkin_cmd.go

```diff
@@ -57,23 +58,30 @@ func (e *CheckinResponse) Validate() error {
 type CheckinCmd struct {
 	client clienter
 	info   agentInfo
+	metaFn metadataFunc
 }
 
 type agentInfo interface {
 	AgentID() string
 }
 
-// NewCheckinCmd creates a new api command.
-func NewCheckinCmd(info agentInfo, client clienter) *CheckinCmd {
+type metadataFunc func() (map[string]interface{}, error)
 
+// NewCheckinCmd creates a new api command.
+func NewCheckinCmd(info agentInfo, client clienter, metaFn metadataFunc) *CheckinCmd {
 	return &CheckinCmd{
 		client: client,
 		info:   info,
+		metaFn: metaFn,
 	}
 }
 
 // Execute enroll the Agent in the Fleet.
 func (e *CheckinCmd) Execute(r *CheckinRequest) (*CheckinResponse, error) {
+	if err := e.injectMetadata(r); err != nil {
```

Comment:
I think is better to just add the metadata to the  `CheckinRequest` Struct instead of delegating the `metadata()` call to the command implementation. Because It's possible that generating metadata cause errors but we might still want to do the call with partial metadata instead of the complete metadata. So if we leave the actual metadata generation outside of the CMD call, the call will only deal with data and not having any behavior logic to do.

Looking at the https://github.com/elastic/kibana/blob/feature-ingest/x-pack/legacy/plugins/fleet/dev_docs/api/agents_checkin.md and the https://github.com/elastic/kibana/blob/feature-ingest/x-pack/legacy/plugins/fleet/dev_docs/api/agents_enroll.md the way we pass metadata appear to be different. @nchaulet Is there a reason why the naming of the `local` metadata is different between both calls? Could we try to make it the same?

Reply:
During checkin you can only update local metadata, it's why the naming of the key is different, but we can make it the same

Reply:
@nchaulet Make sense lets keep it that way.